p2p/net/conn.go

```diff
@@ -159,82 +168,94 @@ func (c *FormattedConnection) Send(m []byte) error {
 
 // Close closes the connection (implements io.Closer). It is go safe.
 func (c *FormattedConnection) Close() {
-	c.closeOnce.Do(func() {
-		close(c.closeChan)
-	})
+	c.wmtx.Lock()
+	if c.closed {
+		c.wmtx.Unlock()
+		return
+	}
+	err := c.close.Close()
+	c.closed = true
+	c.wmtx.Unlock()
+	if err != nil {
+		c.logger.Warning("error while closing with connection %v, err: %v", c.RemotePublicKey().String(), err)
+	}
 }
 
-// Closed Reports whether the connection was closed. It is go safe.
+// Closed returns whether the connection is closed
 func (c *FormattedConnection) Closed() bool {
+	c.wmtx.Lock()
+	defer c.wmtx.Unlock()
 	return c.closed
 }
 
-func (c *FormattedConnection) shutdown(err error) {
-	c.closed = true
-	c.logger.Debug("Shutting down conn with %v err=%v", c.RemotePublicKey().String(), err)
-	if err != ErrConnectionClosed {
-		c.networker.publishClosingConnection(c)
-	}
-	c.formatter.Close()
-}
-
 var ErrTriedToSetupExistingConn = errors.New("tried to setup existing connection")
 var ErrIncomingSessionTimeout = errors.New("timeout waiting for handshake message")
 
 func (c *FormattedConnection) setupIncoming(timeout time.Duration) error {
-	var err error
-	tm := time.NewTimer(timeout)
+	be := make(chan struct {
+		b []byte
+		e error
+	})
+
+	go func() {
```

Comment:
If the other side of this connection never sends a session request, what will terminate this go routine? Isn't it a leak?